src/omnipool/invariants.rs

```diff
@@ -0,0 +1,323 @@
+    ) {
+        let result = calculate_buy_for_hub_asset_state_changes(&asset_out, amount,
+            asset_fee,
+        );
+
+        assert!(result.is_some());
+
+        let state_changes = result.unwrap();
+
+        let asset_out_state = asset_out.clone();
+        let asset_out_state = asset_out_state.delta_update(&state_changes.asset).unwrap();
+        assert_asset_invariant(&asset_out, &asset_out_state,  FixedU128::from((TOLERANCE, ONE)), "Sell update invariant - token out");
+    }
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn buy_update_invariants_with_fees(asset_in in asset_state(), asset_out in asset_state(),
+        amount in trade_amount(),
+        asset_fee in fee(),
+        protocol_fee in fee()
+    ) {
+        let result = calculate_buy_state_changes(&asset_in, &asset_out, amount,
+            asset_fee,
+            protocol_fee,
+            Balance::default()
+        );
+
+        // ignore the invalid result
+        if let Some(state_changes) = result {
+            let asset_in_state = asset_in.clone();
+            let asset_in_state = asset_in_state.delta_update(&state_changes.asset_in).unwrap();
+            assert_asset_invariant(&asset_in, &asset_in_state,  FixedU128::from((TOLERANCE, ONE)), "Buy update invariant - token in");
+
+            let asset_out_state = asset_out.clone();
+            let asset_out_state = asset_out_state.delta_update(&state_changes.asset_out).unwrap();
+            assert_asset_invariant(&asset_out, &asset_out_state,  FixedU128::from((TOLERANCE, ONE)), "Buy update invariant - token out");
+        }
+    }
+}
+#[test]
+fn buy_update_invariants_no_fees_case() {
+    let asset_in = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 10000000000000000,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
+    };
+    let asset_out = AssetReserveState {
+        reserve: 10000000000000000,
+        hub_reserve: 89999999999999991,
+        shares: 10000000000000000,
+        protocol_shares: 10000000000000000,
+        tvl: 10000000000000000,
+    };
+    let amount = 1_000_000_000_000_000;
+
+    let result = calculate_buy_state_changes(
+        &asset_in,
+        &asset_out,
+        amount,
+        Permill::from_percent(0),
+        Permill::from_percent(0),
+        Balance::default(),
+    );
+
+    assert!(result.is_none()); // This fails because of not enough asset out in pool out
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1000))]
+    #[test]
+    fn buy_update_invariants_no_fees(asset_in in asset_state(), asset_out in asset_state(),
+        amount in trade_amount()
+    ) {
+        let result = calculate_buy_state_changes(&asset_in, &asset_out, amount,
+
+        Permill::from_percent(0),
+        Permill::from_percent(0),
+            Balance::default()
+        );
+
+        // ignore the invalid result
```

Comment:
I think that instead of telling what we do not do, would be better to tell what we do

```suggestion
        // do assertions if the result is valid
```

If you go for this, please do it in the previous test as well.